### ISSUE TITLE
Invalid activation info in grain directory?

### ISSUE THREAD

**Issue by Asakuri** [dotnet/orleans#7340]
After one rolling upgrade of our application service, we're seeing continous grain invocation failure for one specific grain instance, other instances of the same grain type don't have this problem, the error message is:

> Intermediate NonExistentActivation for message Transient Rejection (info: Exception while sending message: Orleans.Runtime.Messaging.ConnectionFailedException: Unable to connect to endpoint S10.241.2.63:11111:370607565. See InnerException
 ---> Orleans.Networking.Shared.SocketConnectionException: Unable to connect to 10.241.2.63:11111. Error: ConnectionRefused
   at Orleans.Networking.Shared.SocketConnectionFactory.ConnectAsync(EndPoint endpoint, CancellationToken cancellationToken)
   at Orleans.Runtime.Messaging.ConnectionFactory.ConnectAsync(SiloAddress address, CancellationToken cancellationToken)
   at Orleans.Internal.OrleansTaskExtentions.MakeCancellable[T](Task`1 task, CancellationToken cancellationToken)
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   --- End of inner exception stack trace ---
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   at Orleans.Runtime.Messaging.ConnectionManager.GetConnectionAsync(SiloAddress endpoint)
   at Orleans.Runtime.Messaging.OutboundMessageQueue.<SendMessage>g__SendAsync|10_0(ValueTask`1 c, Message m)) Response S10.241.2.63:11111:370607565*grn/CECB6524/00000000+608a9eb7-f19a-479d-beb8-e0bf87da0b17@195a3387->S10.240.9.79:11111:371983041*cli/8fdffc82@55b92286 #119825, with Exception Orleans.Runtime.Catalog+NonExistentActivationException: Non-existent activation: [ActivationAddress: S10.240.9.79:11111:371983041*cli/8fdffc82@55b92286, Full GrainId: [GrainId: *cli/46a91aaf6aca8a400faab4328fdffc820400000000000000-0x0090C4B2, IdCategory: Client, BaseTypeCode: 0 (x0), PrimaryKey: 6aca8a40-1aaf-46a9-82fc-df8f32b4aa0f (x{0x6aca8a40,0x1aaf,0x46a9,{0x82,0xfc,0xdf,0x8f,0x32,0xb4,0xaa,0x0f}}), UniformHashCode: 9487538 (0x0090C4B2)], Full ActivationId: @47be387a351545a08bc659cf55b922860000000000000000], grain type: .
   at Orleans.Runtime.Catalog.GetOrCreateActivation(ActivationAddress address, Boolean newPlacement, String grainType, String genericArguments, Dictionary`2 requestContextData, Task& activatedPromise)
   at Orleans.Runtime.Dispatcher.ReceiveMessage(Message message)

We're seeing these error logs from all silo logs, it seems they all failed to connect one specific IP address, which most probably is the IP address of an already dead silo. Looks like the activation info is invalid for this grain instance but orleans fails to detect and remove this entry from grain directory?

We're using Orleans 3.3, after looking through the source code, my suspicion is that this is caused by the grain directory handoff during the graceful shutdown, this part of logic is removed in the latest orleans code, but we need to find the root cause for our service failure, is the following case possible to happen and result to the error we met?

1. Grain A is activated in Silo1, the activation info is stored in Silo2's grain directory
2. Silo1 is dead but before this is propogated to Silo2's membership service, Silo2 also begins to shutdown and handoff its grain directory info to Silo3.
3. Silo3 is a newly started silo and receive the grain directory info from Silo2 which contains the activation info of GrainA->Silo1(it seems there's no check for target silo status when handoff manager accepts the handoff partition), but since Silo3 is newly started, Silo1 is already in dead status when it tries to init the membership table, no silo dead event will be triggered therefore this invalid activation entry will no longer be removed, and all the following invocation of GrainA will fail since Silo1 could not be connected?

Is the above assumption possible? If not, what could be the real problem for the error we met? How do we debug further?
Any help is appreciated, thanks a lot in advance.

**Comment by Asakuri** [comment]
> I think your issue is caused by this. How often do you clean the membership table?

```
.Configure<ClusterMembershipOptions>(options =>
                {
                    options.DefunctSiloExpiration = TimeSpan.FromDays(4);
                    options.DefunctSiloCleanupPeriod = TimeSpan.FromHours(8);
                });
```


> If it was marked as dead in the table, you should not see this issue.

That silo should be marked as dead and reside in table for 4 days according to our configuration, I checked the log history, for the first 4 days after this error happened, the rejection message info is "known-dead silo S10.241.2.63:11111:370607565". I checked the log of "silo S10.241.2.63", the log shows that it failed to deactivate this activation when it shut down because the hop limit exceeded. It seems the membership was not stable at that time because of the rolling upgrade and this activation info was left orphaned without cleanup.

> {"ClassName":"Orleans.Runtime.OrleansException","Message":"Silo S10.241.0.119:11111:370607565 is not owner of *grn\/35C0D1D\/00000000+6569, cannot forward UnregisterManyAsync to owner S10.241.2.18:11111:370607635 because hop limit is reached","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"   at Orleans.Runtime.GrainDirectory.LocalGrainDirectory.UnregisterManyAsync(List`1 addresses, UnregistrationCause cause, Int32 hopCount)\n   at Orleans.Runtime.GrainDirectory.DhtGrainLocator.UnregisterExecute()\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2146233088,"Source":"Orleans.Runtime","WatsonBuckets":null}

**Comment by benjaminpetit** [comment]
I think you are right, you might got a poinsoned entry like this

But what you experienced should be fixed now in 3.5.

**Comment by ReubenBond** [comment]
Fixed
